src/card_operations/card_internal.c

```diff
@@ -0,0 +1,370 @@
+ * GLOBAL VARIABLES
+ *****************************************************************************/
+
+/*****************************************************************************
+ * STATIC FUNCTIONS
+ *****************************************************************************/
+static card_error_type_e load_session_key_if_pairing_required(
+    card_operation_data_t *card_data) {
+  if (false == card_data->nfc_data.pairing_required)
+    NFC_RETURN_SUCCESS(card_data);
+
+  card_data->nfc_data.keystore_index =
+      is_paired(card_data->nfc_data.card_key_id);
+
+  if (-1 == card_data->nfc_data.keystore_index) {
+    NFC_RETURN_ERROR_TYPE(card_data, CARD_OPERATION_PAIRING_REQUIRED);
+  }
+
+  const uint8_t *pairing_key =
+      get_keystore_pairing_key(card_data->nfc_data.keystore_index);
+  init_session_keys(pairing_key, pairing_key + 32, NULL);
+  NFC_RETURN_SUCCESS(card_data);
+}
+
+static void select_applet_and_update_tapped_card(
+    NFC_connection_data *nfc_data) {
+  uint8_t temp = 0;
+
+  temp = nfc_data->acceptable_cards;
+  nfc_data->status = nfc_select_applet(nfc_data->family_id,
+                                       &nfc_data->acceptable_cards,
+                                       NULL,
+                                       nfc_data->card_key_id,
+                                       &nfc_data->recovery_mode);
+
+  /* The tapped_card information should be persistent, as it is used at later
+   * stage in the flow For example, in the second half of card-verification,
+   * the card tapped in first half becomes the acceptable card for that half.
+   * It is the knowledge of previous round of card tapping, hence it's update
+   * should happen only here or at the fresh start of a card tap flow. */
+  nfc_data->tapped_card = (temp ^ nfc_data->acceptable_cards);
+}
+
+static card_error_type_e handle_nfc_errors(card_operation_data_t *card_data) {
+  card_data->nfc_data.tapped_card = 0;
+
+  if ((card_data->nfc_data.status == NFC_CARD_ABSENT) ||
+      (nfc_diagnose_card_presence() != 0)) {
+    instruction_scr_change_text(ui_text_card_removed_fast, true);
+    if ((!--card_data->nfc_data.card_absent_retries)) {
+      card_data->nfc_data.status = NFC_CARD_ABSENT;
+      NFC_RETURN_RETAP_ERROR(card_data, ui_text_card_freq_discon_fault);
+    }
+  } else if (!(--card_data->nfc_data.retries)) {
+    NFC_RETURN_ABORT_ERROR(card_data, ui_text_unknown_error_contact_support);
+  } else if ((card_data->nfc_data.status & NFC_ERROR_BASE) == NFC_ERROR_BASE) {
+    instruction_scr_change_text(ui_text_card_align_with_device_screen, true);
+    nfc_deselect_card();
+  }
+
+  // probably comm failure; guess is NFC teardown; reconnect with applet
+  // silently retry to connect; can't connect here, cards_state is important
+  NFC_RETURN_SUCCESS(card_data);
+}
+
+static card_error_type_e handle_wait_for_card_selection(
+    card_operation_data_t *card_data) {
+  evt_status_t evt_status = {0};
+
+  while (1) {
```

Comment:
cross-check: is this while needed?

my concern comes from lots of nested whiles appearing here.
additionally, add the expected behavior of this function in doxygen comments. something around covering following thoughts:

will it return on aborts? 
will it ignore inactivity events?

Reply:
Reason for keeping the while loop is to handle deselection of card happening in `get_events`.

Updated the handling as discussed.

Reply:
Added comment and updated as discussed